src/iterators/dfs.jl

```diff
@@ -0,0 +1,83 @@
+"""
+    DFSIterator
+
+`DFSIterator` is a subtype of [`VertexIterator`](@ref) to iterate through graph vertices using a depth-first search. A source node(s) is optionally supplied as an `Int` or an array-like type that can be indexed if supplying multiple sources. If no source is provided, it defaults to the first vertex.
+    
+# Examples
+```julia-repl
+julia> g = smallgraph(:house)
+{5, 6} undirected simple Int64 graph
+
+julia> for node in DFSIterator(g, 3)
+           display(node)
+       end
+1
+2
+4
+3
+5
+```
+"""
+struct DFSIterator{S,G<:AbstractGraph}
+    graph::G
+    source::S
+end
+
+"""
+    DFSVertexIteratorState
+
+`DFSVertexIteratorState` is a struct to hold the current state of iteration
+in DFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct DFSVertexIteratorState
+    visited::BitVector
+    queue::Vector{Int}
+end
+
+DFSIterator(g::AbstractGraph) = DFSIterator(g, first(vertices(g)))
+
+Base.length(t::DFSIterator) = nv(t.graph)
```

Comment:
I added a known length but on second thought it may not work for unconnected graphs. In which case the termination criteria in the iteration also need to change

Reply:
same as before